### PR TITLE
docs: add some more information about dag-pb, show protobuf decoding

### PR DESCRIPTION
**Pull request by mrd0ll4r**
I discovered a few things about IPFS data types which I thought might be useful to others.

### CODE REVIEW

File: docs/ipfs_datatypes.md [hunk 1]
```diff
@@ -53,8 +53,8 @@ Reference: https://ipld.io/
 In IPFS, all CIDs refer to objects with the following properties:
 
 * `cid`: the identifier (CIDv0 or CIDv1, as described above).
-* `size`: the cumulative size of the object, including the sizeo of the objects linked to.
-* `links`: a list of links to other objects in CID format.
+* `size`: the cumulative size of the object, including the size of the objects linked to.
+* `links`: a list of links to other objects in CID format, except for `raw` objects.
```
Comment (dokterbob) [inline docs/ipfs_datatypes.md:57]: Raw objects can't have links? Did know that!

File: docs/ipfs_datatypes.md [hunk 2]
```diff
@@ -63,15 +63,19 @@ Objects may be explored with the IPLD explorer: https://explore.ipld.io/
 
 Within IPLD, the following data formats are currently usage within IPFS:
 
-1. `dag-pb`: UnixFS protobuf format for files, raw data, HAMT and directories.
+1. `dag-pb`: Protobuf wrapper around UnixFS protobuf format for files, raw data, HAMT, and directories.
```
Comment (dokterbob) [inline docs/ipfs_datatypes.md:66]: Protobuf wrapper around protobuf format? Double protobuf?

File: docs/ipfs_datatypes.md [hunk 3]
```diff
@@ -299,6 +303,42 @@ $ ipfs --encoding=json object stat /ipfs/QmWyDJmrr6cRwEpTF2VGhWDi4uytrDHT8S5BptV
...
+    `02` is the value in `varint` encoding, which is just `2` in decimal.
+    This is the `Type` field in the [unixfs protobuf spec](https://github.com/ipfs/go-unixfs/blob/master/pb/unixfs.proto).
+    The type is `file`.
+- `12c308` -> `12` is `1 0010`, which tells us this is field number `2`, type `010`, which is a `bytes` in this case.
+    `c308` is the length, as `varint`: `1100 0011 0000 1000` -> `100 0011 000 1000` -> reverse to `000 1000 100 0011` -> `1091` in decimal.
+- some data (which is actually `1091` bytes long, nice!)
+- `18c3 08` -> `18` is `1 1000`, which tells us this is field number `3`, type `000`, which is a `varint` in this case.
+    `c308` is, again `1091` in decimal, which is good, because this is a file and that's what you'd expect.
+    This is the `filesize` field.
+
```
Comment (dokterbob) [inline docs/ipfs_datatypes.md:341]: \o/

File: docs/ipfs_datatypes.md (hunk 2 above)
Comment (mrd0ll4r) [reply docs/ipfs_datatypes.md:66]: Yeah, the `Data` field of the [dag-pb protobuf](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-pb.md) holds the UnixFS protobuf. And it seems like this is always the case. I was quite puzzled too, which is why I went through the hassle of manually decoding the blocks and trying to dig up the dag-pb spec 🙃 

Sounds a bit weird though, I'm open to suggestions making it sound nicer!

**Comment by dokterbob** [comment]
Well, the way you express it here, including references, sheds a bit more light and in my eyes would suffice.

I feel, at times, that IPFS would have done well to do a full rewrite from scratch after the proof of concept, instead now there’s a lot of legacy like this… ;)

**Comment by mrd0ll4r** [comment]
I added the information from my comment into the text to make things clearer :)

**Comment by dokterbob** [comment]
Great, thanks!

**Pull request merged**